agent/consul/leader_peering.go

```diff
@@ -27,8 +29,83 @@ import (
 	"github.com/hashicorp/consul/proto/pbpeerstream"
 )
 
+var leaderExportedServicesCountKey = []string{"consul", "peering", "exported_services"}
+var LeaderPeeringMetrics = []prometheus.GaugeDefinition{
+	{
+		Name: leaderExportedServicesCountKey,
+		Help: "A gauge that tracks how many services are exported for the peering. " +
+			"The labels are \"peering\" and, for enterprise, \"partition\". " +
+			"We emit this metric every 9 seconds",
+	},
+}
+
 func (s *Server) startPeeringStreamSync(ctx context.Context) {
 	s.leaderRoutineManager.Start(ctx, peeringStreamsRoutineName, s.runPeeringSync)
+	s.leaderRoutineManager.Start(ctx, peeringStreamsMetricsRoutineName, s.runPeeringMetrics)
+}
+
+func (s *Server) runPeeringMetrics(ctx context.Context) error {
+	ticker := time.NewTicker(s.config.MetricsReportingInterval)
+	defer ticker.Stop()
+
+	logger := s.logger.Named(logging.PeeringMetrics)
+	defaultMetrics := metrics.Default
+
+	retryLoopBackoff(ctx, func() error {
```

Comment:
This retry loop shouldn't be necessary if we're ticking at a set interval and errors are suppressed, right?

Reply:
actually, we may error out on the `PeeringList` call for instance

Reply:
The error eventually gets captured by `retryLoopBackoff` and logged anyways so let's remove the retry loop and log it directly.

It's very unlikely for `PeeringList` to fail and if it does, other parts of Consul are probably also failing.

The issue with the retry loop is that you're no longer emitting metrics at a regular interval; if hypothetically something was wrong, we would want the error logs to happen regularly instead of backing off exponentially.

Reply:
hm the cost of not having metrics is high for me but I don't like not having the metric on a regular cadence as you are pointing out;

let me try something; I'll remove the `retryLoopBackoff` and isolate the metric emission to a `emitOnce` func